shachain/element.go

```diff
@@ -0,0 +1,134 @@
+package shachain
+
+import (
+	"crypto/sha256"
+	"errors"
+
+	"github.com/roasbeef/btcd/wire"
+)
+
+// Element represent the entityt which contains the hash and
+// correponding to it index. By comparing two indexes we may change hash in
+// such way to derive another element.
+type Element struct {
+	index index
+	hash  *wire.ShaHash
```

Comment:
`hash` should be a `wire.ShaHash` rather than a `*wire.ShaHash`. Doing so provides better cache locality, and can possibly allow for items to be GC'd sooner. I think this outweighs the cost of copying 32 bytes.